x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/RegisterAnalyzeAction.java

```diff
@@ -0,0 +1,217 @@
+            if (repository instanceof BlobStoreRepository == false) {
+                throw new IllegalArgumentException("repository [" + request.getRepositoryName() + "] is not a blob-store repository");
+            }
+            if (repository.isReadOnly()) {
+                throw new IllegalArgumentException("repository [" + request.getRepositoryName() + "] is read-only");
+            }
+            final BlobStoreRepository blobStoreRepository = (BlobStoreRepository) repository;
+            final BlobPath path = blobStoreRepository.basePath().add(request.getContainerPath());
+            final BlobContainer blobContainer = blobStoreRepository.blobStore().blobContainer(path);
+
+            logger.trace("handling [{}]", request);
+
+            assert task instanceof CancellableTask;
+
+            ActionListener.run(listener.<Void>map(ignored -> ActionResponse.Empty.INSTANCE), l -> {
+                final String registerName = request.getRegisterName();
+                long initialValue;
+                try {
+                    initialValue = blobContainer.getRegister(registerName);
+                } catch (ConcurrentRegisterOperationException e) {
+                    // Concurrent activity prevents us from even reading the initial value, but this is just a best-effort thing so we can
+                    // proceed anyway.
+                    initialValue = 0;
+                } catch (UnsupportedOperationException e) {
+                    // Registers are not supported on all repository types, and that's ok. If it's not supported here then the final check
+                    // will also be unsupported, so it doesn't matter that we didn't do anything before this successful response.
+                    l.onResponse(null);
+                    return;
+                }
+
+                if (initialValue < 0 || initialValue >= request.getRequestCount()) {
+                    throw new IllegalStateException("register holds unexpected value [" + initialValue + "]");
+                }
+
+                class Execution extends ActionRunnable<Void> {
+                    private long currentValue;
+
+                    Execution(long currentValue) {
+                        super(l);
+                        this.currentValue = currentValue;
+                    }
+
+                    @Override
+                    protected void doRun() throws Exception {
+                        if (((CancellableTask) task).notifyIfCancelled(listener)) {
+                            return;
+                        }
+
+                        try {
+                            final var witness = blobContainer.compareAndExchangeRegister(registerName, currentValue, currentValue + 1);
+                            if (witness == currentValue) {
+                                listener.onResponse(null);
+                            } else if (witness < currentValue || witness >= request.getRequestCount()) {
```

Comment:
`witness >= request.getRequestCount()` I think this is a valid scenario? i.e. another task runs > `request.getRequestCount()` updates while this task is waiting to be executed. Am I missing something here?

Reply:
There are only `request.getRequestCount()` increment requests in total, and this one hasn't succeeded yet, so the current value must be smaller.

Reply:
Right, for some reason I thought that the variable was created within the loop in https://github.com/elastic/elasticsearch/pull/93955/files#diff-2685c0b9e390123278fc8429f89d237daf9ed0d74ba41fd9ccc8318d077ca69cR454-R465 🤦